### DISCUSSION TITLE
How the heck do i test if the bot has a certain permission in message event

### DISCUSSION THREAD

**Discussion by MCorange99** [serenity-rs/serenity#2497] [Q&A] [answered]
How the heck do i test if the bot has a certain permission in message event.
Cause it returns none.
Code im using:

```rust
msg.member.clone().unwrap().permissions.unwrap(); // panics cause permissions is none
```

**Comment by arqunis** [comment, marked as answer]
The [`Member::permissions`](https://serenity-rs.github.io/serenity/current/serenity/model/prelude/struct.Member.html#structfield.permissions) field is only `Some` when the `Member` instance is from an interaction, like a slash command or message component. In regular situations, you have to calculate the permissions the member has.

There are plenty of methods for doing that, so I recommend using the one that's most relevant to you:
1. Permissions based on roles only:
   - https://serenity-rs.github.io/serenity/current/serenity/model/guild/struct.Member.html#method.permissions (note: this is the method, not the field!)
   - https://serenity-rs.github.io/serenity/current/serenity/model/guild/struct.PartialGuild.html#method.member_permissions
   - https://serenity-rs.github.io/serenity/current/serenity/model/guild/struct.Guild.html#method.member_permissions
2. Permissions based on roles and channel overwrites:
   - https://serenity-rs.github.io/serenity/current/serenity/model/guild/struct.PartialGuild.html#method.user_permissions_in
   - https://serenity-rs.github.io/serenity/current/serenity/model/guild/struct.Guild.html#method.user_permissions_in
   - https://serenity-rs.github.io/serenity/current/serenity/model/channel/struct.GuildChannel.html#method.permissions_for_user
